File: advanced data structures/segment trees/No-Inversions.java
import java.io.*;
import java.util.*;

public class Main {

    static class Node {
        long cnt, lcnt, rcnt;
        int left, right;
        boolean isGood;

        Node() {
            cnt = lcnt = rcnt = 0;
            left = right = 0;
            isGood = true;
        }

        Node(int u) {
            cnt = lcnt = rcnt = (u > 0 ? 1 : 0);
            left = right = u;
            isGood = true;
        }

        Node merge(Node a) { // a is right node
            Node f = new Node();
            f.cnt = this.cnt + a.cnt;

            if (this.right != 0 && a.left != 0 && this.right <= a.left) {
                f.cnt += this.rcnt * a.lcnt;
                f.isGood = this.isGood && a.isGood;
                f.lcnt = (this.isGood ? 1 : 0) * a.lcnt;
                f.rcnt = (a.isGood ? 1 : 0) * this.rcnt;
            } else {
                f.isGood = false;
            }

            f.lcnt += this.lcnt;
            f.rcnt += a.rcnt;
            f.left = this.left;
            f.right = a.right;

            return f;
        }

        void assign(int u) {
            cnt = lcnt = rcnt = (u > 0 ? 1 : 0);
            left = right = u;
            isGood = true;
        }
    }

    static class SegmentTree {
        int n, N;
        Node[] seg;

        SegmentTree(int n) {
            this.n = n;
            this.N = 4 * n;
            seg = new Node[N + 5];
            for (int i = 0; i < seg.length; i++) seg[i] = new Node();
        }

        void update(int x, int s, int e, int p, int u) {
            if (e < s || s > p || e < p) return;

            if (s == e) {
                seg[x].assign(u);
                return;
            }

            int m = (s + e) >> 1;
            update(x << 1, s, m, p, u);
            update(x << 1 | 1, m + 1, e, p, u);

            seg[x] = seg[x << 1].merge(seg[x << 1 | 1]);
        }

        Node query(int x, int s, int e, int l, int r) {
            if (l > e || r < s || s > e) return new Node();

            if (l <= s && e <= r) return seg[x];

            int m = (s + e) >> 1;
            Node left = query(x << 1, s, m, l, r);
            Node right = query(x << 1 | 1, m + 1, e, l, r);

            return left.merge(right);
        }

        Node query(int l, int r) {
            if (l > r) return new Node();
            return query(1, 1, n, l, r);
        }

        void update(int p, int u) {
            update(1, 1, n, p, u);
        }
    }

    // Fast IO
    static class FastScanner {
        BufferedReader br;
        StringTokenizer st;

        FastScanner() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        String next() throws IOException {
            while (st == null || !st.hasMoreElements())
                st = new StringTokenizer(br.readLine());
            return st.nextToken();
        }

        int nextInt() throws IOException {
            return Integer.parseInt(next());
        }
    }

    public static void solve(FastScanner fs) throws Exception {
        int n = fs.nextInt();
        String s = fs.next();

        SegmentTree segtree = new SegmentTree(n);

        for (int i = 0; i < n; i++) {
            segtree.update(i + 1, s.charAt(i) - 'a' + 1);
        }

        int q = fs.nextInt();

        StringBuilder out = new StringBuilder();

        while (q-- > 0) {
            int x = fs.nextInt();
            int y = fs.nextInt();
            out.append(segtree.query(x, y).cnt).append('\n');
        }

        System.out.print(out);
    }

    public static void main(String[] args) throws Exception {
        FastScanner fs = new FastScanner();
        int test = fs.nextInt();
        while (test-- > 0) solve(fs);
    }
}

File: advanced data structures/segment trees/no inversions.swift
import Foundation

struct Node {
    var cnt: Int64 = 0
    var lcnt: Int64 = 0
    var rcnt: Int64 = 0
    var left: Int = 0
    var right: Int = 0
    var isGood: Bool = true

    init() {}

    init(_ u: Int) {
        let val: Int64 = (u > 0 ? 1 : 0)
        cnt = val
        lcnt = val
        rcnt = val
        left = u
        right = u
        isGood = true
    }

    static func merge(_ a: Node, _ b: Node) -> Node {
        if a.left == 0 { return b }
        if b.left == 0 { return a }

        var f = Node()
        f.cnt = a.cnt + b.cnt

        if a.right <= b.left {
            f.cnt += a.rcnt * b.lcnt
            f.isGood = a.isGood && b.isGood
            f.lcnt = (a.isGood ? 1 : 0) * b.lcnt
            f.rcnt = (b.isGood ? 1 : 0) * a.rcnt
        } else {
            f.isGood = false
        }

        f.lcnt += a.lcnt
        f.rcnt += b.rcnt
        f.left = a.left
        f.right = b.right

        return f
    }
}

final class SegTree {
    let n: Int
    var tree: [Node]

    init(_ s: [Int]) {
        n = s.count
        tree = Array(repeating: Node(), count: 2 * n)

        // build leaves
        for i in 0..<n {
            tree[n + i] = Node(s[i])
        }

        // build internal nodes
        for i in stride(from: n - 1, through: 1, by: -1) {
            tree[i] = Node.merge(tree[i << 1], tree[i << 1 | 1])
        }
    }

    func query(_ l: Int, _ r: Int) -> Node {
        var l = l + n - 1
        var r = r + n - 1

        var leftRes = Node()
        var rightRes = Node()

        while l <= r {
            if (l & 1) == 1 {
                leftRes = Node.merge(leftRes, tree[l])
                l += 1
            }
            if (r & 1) == 0 {
                rightRes = Node.merge(tree[r], rightRes)
                r -= 1
            }
            l >>= 1
            r >>= 1
        }

        return Node.merge(leftRes, rightRes)
    }
}

// Fast IO
final class FastInput {
    private var data: [String] = []
    private var idx = 0

    init() {
        while let line = readLine() {
            data += line.split(separator: " ").map { String($0) }
        }
    }

    func nextInt() -> Int {
        defer { idx += 1 }
        return Int(data[idx])!
    }

    func next() -> String {
        defer { idx += 1 }
        return data[idx]
    }
}

// Main
let input = FastInput()
let t = input.nextInt()

var output = String()

for _ in 0..<t {
    let n = input.nextInt()
    let s = Array(input.next())

    var arr = [Int](repeating: 0, count: n)
    for i in 0..<n {
        arr[i] = Int(s[i].asciiValue! - 96)
    }

    let st = SegTree(arr)

    let q = input.nextInt()

    for _ in 0..<q {
        let l = input.nextInt()
        let r = input.nextInt()
        output += "\(st.query(l, r).cnt)\n"
    }
}

print(output)
